molr-mole-core/src/test/java/io/molr/mole/core/tree/RunnableLeafsMoIeLoopIntegrationTest.java

```diff
@@ -0,0 +1,114 @@
+package io.molr.mole.core.tree;
+
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.assertj.core.api.Assertions;
+import org.junit.Test;
+
+import com.google.common.collect.Sets;
+
+import io.molr.commons.domain.Block;
+import io.molr.commons.domain.ExecutionStrategy;
+import io.molr.commons.domain.ListOfStrings;
+import io.molr.commons.domain.Mission;
+import io.molr.commons.domain.MissionHandle;
+import io.molr.commons.domain.MissionOutput;
+import io.molr.commons.domain.MissionRepresentation;
+import io.molr.commons.domain.Placeholder;
+import io.molr.commons.domain.Placeholders;
+import io.molr.commons.domain.StrandCommand;
+import io.molr.mole.core.api.Mole;
+import io.molr.mole.core.runnable.RunnableLeafsMission;
+import io.molr.mole.core.runnable.RunnableLeafsMole;
+import io.molr.mole.core.runnable.lang.RunnableLeafsMissionSupport;
+
+/**
+ * @author krepp
+ */
+public class RunnableLeafsMoIeLoopIntegrationTest {
+
+    private final static Placeholder<String> ITEM_PLACEHOLDER = Placeholder.aString("aForEachLoop.deviceName");
+    private final static List<String> ITEM_LIST = new ArrayList<>(Arrays.asList("A", "B", "C"));
+    private final static List<String> ITEM_LIST_2 = new ArrayList<>(Arrays.asList("D", "E", "F"));
+    private final static String PARAMETER_NAME_DEVICE_NAMES = "deviceNames";
+    private final static String PARAMETER_NAME_DEVICE_NAMES_2 = "deviceNames2";
+    
+    RunnableLeafsMission mission() {
+
+        return new RunnableLeafsMissionSupport() {
+            {
+
+                Placeholder<ListOfStrings> collectionPlaceholder = mandatory(Placeholder.aListOfStrings(PARAMETER_NAME_DEVICE_NAMES));
+
+                optional(Placeholders.EXECUTION_STRATEGY, ExecutionStrategy.ABORT_ON_ERROR.name());
+                                
+                root("root1").sequential().as(missionRoot -> {// 0
+                    
+                    missionRoot.leafForEach("aForEachLoop", collectionPlaceholder, ITEM_PLACEHOLDER, (in, out) -> {
```

Comment:
Next thing: Having a foreach placeholder defined upfront is potentially not the right thing ... (would be as if having to define the loop variable outside the loop). I think this should be implicit.

Reply:
And: Never have more than one (or maxmax 2 ) parameters per method.

Reply:
Having all this said... I could see something like this as a proposal:
```java
Placeholder<Iterable<T>> collectionPlaceholder;

b.branch("foreach device").forEach(collectionPlaceholder).parallel().as(b1 ->{
   // b1 would be of type ForEachBranch ... take inspiration from Contextual branch
    b1.branch("another subbranch").as(...);
    b1.leaf("Switch on").run(e -> {
        // e being the loop variable
    });
  });
```

The run methods would have corresponding overloads ... e.g. 
```
run((T e, In in , Out out) -> {});
```

It could be, that the value of the Loop variable would be transported within the In object. If we would only allow one loop (non-nexted), we could apply a similar pattern than with the context: just define one global constant for a placeholder (e.g. loopVariable())... 

If we would allow nested loops .. things get a bit more tricky ... So probably best to go for one level for at first ..?

Reply:
Next discussion roung:

```java
 Placeholder<ListOfStrings> collectionPlaceholder = mandatory(Placeholder.aListOfStrings("deviceNames"));
                Placeholder<String> itemPlaceholder = Placeholder.aString("aForEachLoop.deviceName");
                optional(Placeholders.EXECUTION_STRATEGY, ExecutionStrategy.ABORT_ON_ERROR.name());
                
                root("root1").sequential().as(missionRoot -> {// 0
                    
                    missionRoot.leafForEach("aForEachLoop", collectionPlaceholder, itemPlaceholder, (in, out) -> {
                        String deviceName = in.get(itemPlaceholder);
                        System.out.println("deviceName: " + deviceName);
                    });
                    
                    missionRoot.branch("main1").parallel().as(main1Root -> {// 1

                        main1Root.leaf("main1Sub1").run(() -> {// 2
                            System.out.println("main1Sub1");
                        });

                        main1Root.branch("main1Sub2").sequential().as(main1Sub2 -> {// 3

                            main1Sub2.leaf("main1Sub2Sub1").run(() -> {// 4
                                System.out.println("main1Sub2Task1");
                            });

                            main1Sub2.leaf("main1Sub2Sub2").run(() -> {// 5
                                System.out.println("main1Sub2Task2");
                            });

                            main1Sub2.leaf("main1Sub2Sub3").run(() -> {// 6
                                System.out.println("main1Sub2Task3");
                            });
                        });

                        main1Root.leaf("main1Sub3").run(() -> {// 7
                            System.out.println("main1Sub1");
                        });
                    });

                    missionRoot.leaf("main2").run((in, out) -> {// 8
                        System.out.println("hello1");
                    });
                    missionRoot.leaf("main3").run((in, out) -> {// 9
                        System.out.println("hello2");
                    });
                });


```